File: hail/python/hail/ir/blockmatrix_writer.py
import abc
import json

from ..typecheck import typecheck_method, sequenceof, nullable, enumeration
from ..expr.types import tvoid, tstr
from ..utils.misc import escape_str


class BlockMatrixWriter(object):
    @abc.abstractmethod
    def render(self):
        pass

    @abc.abstractmethod
    def _type(self):
        pass

    @abc.abstractmethod
    def __eq__(self, other):
        pass


class BlockMatrixNativeWriter(BlockMatrixWriter):
    @typecheck_method(path=str, overwrite=bool, force_row_major=bool, stage_locally=bool)
    def __init__(self, path, overwrite, force_row_major, stage_locally):
        self.path = path
        self.overwrite = overwrite
        self.force_row_major = force_row_major
        self.stage_locally = stage_locally

    def render(self):
        writer = {'name': 'BlockMatrixNativeWriter',
                  'path': self.path,
                  'overwrite': self.overwrite,
                  'forceRowMajor': self.force_row_major,
                  'stageLocally': self.stage_locally}
        return escape_str(json.dumps(writer))

    def _type(self):
        return tvoid

    def __eq__(self, other):
        return isinstance(other, BlockMatrixNativeWriter) and \
            self.path == other.path and \
            self.overwrite == other.overwrite and \
            self.force_row_major == other.force_row_major and \
            self.stage_locally == other.stage_locally


class BlockMatrixBinaryWriter(BlockMatrixWriter):
    @typecheck_method(path=str)
    def __init__(self, path):
        self.path = path

    def render(self):
        writer = {'name': 'BlockMatrixBinaryWriter',
                  'path': self.path}
        return escape_str(json.dumps(writer))

    def _type(self):
        return tstr

    def __eq__(self, other):
        return isinstance(other, BlockMatrixBinaryWriter) and \
            self.path == other.path


class BlockMatrixRectanglesWriter(BlockMatrixWriter):
    @typecheck_method(path=str,
                      rectangles=sequenceof(sequenceof(int)),
                      delimiter=str,
                      binary=bool)
    def __init__(self, path, rectangles, delimiter, binary):
        self.path = path
        self.rectangles = rectangles
        self.delimiter = delimiter
        self.binary = binary

    def render(self):
        writer = {'name': 'BlockMatrixRectanglesWriter',
                  'path': self.path,
                  'rectangles': self.rectangles,
                  'delimiter': self.delimiter,
                  'binary': self.binary}
        return escape_str(json.dumps(writer))

    def _type(self):
        return tvoid

    def __eq__(self, other):
        return isinstance(other, BlockMatrixRectanglesWriter) and \
            self.path == other.path and \
            self.rectangles == other.rectangles and \
            self.delimiter == other.delimiter and \
            self.binary == other.binary


class BlockMatrixMultiWriter(object):
    @abc.abstractmethod
    def render(self):
        pass

    def _type(self):
        return tvoid

    @abc.abstractmethod
    def __eq__(self, other):
        pass


class BlockMatrixBinaryMultiWriter(BlockMatrixMultiWriter):
    @typecheck_method(prefix=str, overwrite=bool)
    def __init__(self, prefix, overwrite):
        self.prefix = prefix
        self.overwrite = overwrite

    def render(self):
        writer = {'name': 'BlockMatrixBinaryMultiWriter',
                  'prefix': self.prefix,
                  'overwrite': self.overwrite}
        return escape_str(json.dumps(writer))

    def _type(self):
        return tvoid

    def __eq__(self, other):
        return isinstance(other, BlockMatrixBinaryMultiWriter) and \
            self.prefix == other.prefix and \
            self.overwrite == other.overwrite


class BlockMatrixTextMultiWriter(BlockMatrixMultiWriter):
    @typecheck_method(prefix=str, overwrite=bool, delimiter=str, header=nullable(str), add_index=bool,
                      compression=nullable(enumeration('gz', 'bgz')), custom_filenames=nullable(sequenceof(str)))
    def __init__(self, prefix, overwrite, delimiter, header, add_index, compression, custom_filenames):
        self.prefix = prefix
        self.overwrite = overwrite
        self.delimiter = delimiter
        self.header = header
        self.add_index = add_index
        self.compression = compression
        self.custom_filenames = custom_filenames

    def render(self):
        writer = {'name': 'BlockMatrixTextMultiWriter',
                  'prefix': self.prefix,
                  'overwrite': self.overwrite,
                  'delimiter': self.delimiter,
                  'header': self.header,
                  'addIndex': self.add_index,
                  'compression': self.compression,
                  'customFilenames': self.custom_filenames}
        return escape_str(json.dumps(writer))

    def _type(self):
        return tvoid

    def __eq__(self, other):
        return isinstance(other, BlockMatrixTextMultiWriter) and \
            self.prefix == other.prefix and \
            self.overwrite == other.overwrite and \
            self.delimiter == other.overwrite and \
            self.header == other.header and \
            self.add_index == other.add_index and \
            self.compression == other.compression and \
            self.custom_filenames == other.custom_filenames


class BlockMatrixPersistWriter(BlockMatrixWriter):
    @typecheck_method(id=str, storage_level=str)
    def __init__(self, id, storage_level):
        self.id = id
        self.storage_level = storage_level

    def render(self):
        writer = {'name': 'BlockMatrixPersistWriter',
                  'id': self.id,
                  'storageLevel': self.storage_level}
        return escape_str(json.dumps(writer))

    def _type(self):
        return tvoid

    def __eq__(self, other):
        return isinstance(other, BlockMatrixPersistWriter) and \
            self.id == other.id and \
            self.storage_level == other.storage_level


class BlockMatrixNativeMultiWriter(BlockMatrixMultiWriter):
    @typecheck_method(prefix=str, overwrite=bool, force_row_major=bool, stage_locally=bool)
    def __init__(self, prefix, overwrite, force_row_major, stage_locally):
        self.prefix = prefix
        self.overwrite = overwrite
        self.force_row_major = force_row_major
        self.stage_locally = stage_locally

    def render(self):
        writer = {'name': 'BlockMatrixNativeMultiWriter',
                  'prefix': self.prefix,
                  'overwrite': self.overwrite,
                  'forceRowMajor': self.force_row_major,
                  'stageLocally': self.stage_locally}
        return escape_str(json.dumps(writer))

    def _type(self):
        return tvoid

    def __eq__(self, other):
        return isinstance(other, BlockMatrixNativeMultiWriter) and \
            self.prefix == other.prefix and \
            self.overwrite == other.overwrite and \
            self.force_row_major == other.force_row_major and \
            self.stage_locally == other.stage_locally

File: hail/src/main/scala/is/hail/expr/ir/BlockMatrixWriter.scala
package is.hail.expr.ir

import is.hail.HailContext
import is.hail.annotations.Region
import is.hail.asm4s._
import is.hail.backend.ExecuteContext
import is.hail.expr.Nat
import is.hail.expr.ir.lowering.{BlockMatrixStage2, LowererUnsupportedOperation}
import is.hail.io.{StreamBufferSpec, TypedCodecSpec}
import is.hail.io.fs.FS
import is.hail.linalg.{BlockMatrix, BlockMatrixMetadata}
import is.hail.types.{BlockMatrixType, TypeWithRequiredness}
import is.hail.types.encoded.{EBlockMatrixNDArray, ENumpyBinaryNDArray, EType}
import is.hail.types.virtual._
import is.hail.utils._
import is.hail.utils.richUtils.RichDenseMatrixDouble

import org.json4s.{jackson, DefaultFormats, Formats, ShortTypeHints}

import java.io.DataOutputStream

object BlockMatrixWriter {
  implicit val formats: Formats = new DefaultFormats() {
    override val typeHints = ShortTypeHints(
      List(
        classOf[BlockMatrixNativeWriter],
        classOf[BlockMatrixBinaryWriter],
        classOf[BlockMatrixRectanglesWriter],
        classOf[BlockMatrixBinaryMultiWriter],
        classOf[BlockMatrixTextMultiWriter],
        classOf[BlockMatrixPersistWriter],
        classOf[BlockMatrixNativeMultiWriter],
      ),
      typeHintFieldName = "name",
    )
  }
}

abstract class BlockMatrixWriter {
  def pathOpt: Option[String]
  def apply(ctx: ExecuteContext, bm: BlockMatrix): Any
  def loweredTyp: Type

  def lower(
    ctx: ExecuteContext,
    s: BlockMatrixStage2,
    evalCtx: IRBuilder,
    eltR: TypeWithRequiredness,
  ): IR =
    throw new LowererUnsupportedOperation(s"unimplemented writer: \n${this.getClass}")
}

case class BlockMatrixNativeWriter(
  path: String,
  overwrite: Boolean,
  forceRowMajor: Boolean,
  stageLocally: Boolean,
) extends BlockMatrixWriter {
  def pathOpt: Option[String] = Some(path)

  def apply(ctx: ExecuteContext, bm: BlockMatrix): Unit =
    bm.write(ctx, path, overwrite, forceRowMajor, stageLocally)

  def loweredTyp: Type = TVoid

  override def lower(
    ctx: ExecuteContext,
    s: BlockMatrixStage2,
    evalCtx: IRBuilder,
    eltR: TypeWithRequiredness,
  ): IR = {
    val etype = EBlockMatrixNDArray(
      EType.fromTypeAndAnalysis(s.typ.elementType, eltR),
      encodeRowMajor = forceRowMajor,
      required = true,
    )
    val spec = TypedCodecSpec(etype, TNDArray(s.typ.elementType, Nat(2)), BlockMatrix.bufferSpec)
    val writer = ETypeValueWriter(spec)

    val paths = s.collectBlocks(evalCtx, "block_matrix_native_writer") { (_, idx, block) =>
      val suffix = strConcat("parts/part-", idx, UUID4())
      val filepath = strConcat(s"$path/", suffix)
      WriteValue(
        block,
        filepath,
        writer,
        if (stageLocally) Some(strConcat(s"${ctx.localTmpdir}/", suffix)) else None,
      )
    }
    RelationalWriter.scoped(path, overwrite, None)(WriteMetadata(
      paths,
      BlockMatrixNativeMetadataWriter(path, stageLocally, s.typ),
    ))
  }
}

case class BlockMatrixNativeMetadataWriter(
  path: String,
  stageLocally: Boolean,
  typ: BlockMatrixType,
) extends MetadataWriter {

  case class BMMetadataHelper(
    path: String,
    blockSize: Int,
    nRows: Long,
    nCols: Long,
    partIdxToBlockIdx: Option[IndexedSeq[Int]],
  ) {
    def write(fs: FS, rawPartFiles: Array[String]): Unit = {
      val partFiles = rawPartFiles.map(_.split('/').last)
      using(new DataOutputStream(fs.create(s"$path/metadata.json"))) { os =>
        implicit val formats = defaultJSONFormats
        jackson.Serialization.write(
          BlockMatrixMetadata(blockSize, nRows, nCols, partIdxToBlockIdx, partFiles),
          os,
        )
      }
      val nBlocks = partIdxToBlockIdx.map(_.length).getOrElse {
        BlockMatrixType.numBlocks(nRows, blockSize) * BlockMatrixType.numBlocks(nCols, blockSize)
      }
      assert(nBlocks == partFiles.length, s"$nBlocks vs ${partFiles.mkString(", ")}")

      info(s"wrote matrix with $nRows ${plural(nRows, "row")} " +
        s"and $nCols ${plural(nCols, "column")} " +
        s"as $nBlocks ${plural(nBlocks, "block")} " +
        s"of size $blockSize to $path")
    }
  }

  def annotationType: Type = TArray(TString)

  def writeMetadata(
    writeAnnotations: => IEmitCode,
    cb: EmitCodeBuilder,
    region: Value[Region],
  ): Unit = {
    val metaHelper =
      BMMetadataHelper(path, typ.blockSize, typ.nRows, typ.nCols, typ.linearizedDefinedBlocks)

    val pc = writeAnnotations.get(cb, "write annotations can't be missing!").asIndexable
    val partFiles = cb.newLocal[Array[String]]("partFiles")
    val n = cb.newLocal[Int]("n", pc.loadLength())
    val i = cb.newLocal[Int]("i", 0)
    cb.assign(partFiles, Code.newArray[String](n))
    cb.while_(
      i < n, {
        val s = pc.loadElement(cb, i).get(cb, "file name can't be missing!").asString
        cb += partFiles.update(i, s.loadString(cb))
        cb.assign(i, i + 1)
      },
    )
    cb += cb.emb.getObject(metaHelper).invoke[FS, Array[String], Unit](
      "write",
      cb.emb.getFS,
      partFiles,
    )
  }

  def loweredTyp: Type = TVoid
}

case class BlockMatrixBinaryWriter(path: String) extends BlockMatrixWriter {
  def pathOpt: Option[String] = Some(path)

  def apply(ctx: ExecuteContext, bm: BlockMatrix): String = {
    RichDenseMatrixDouble.exportToDoubles(ctx.fs, path, bm.toBreezeMatrix(), forceRowMajor = true)
    path
  }

  def loweredTyp: Type = TString

  override def lower(
    ctx: ExecuteContext,
    s: BlockMatrixStage2,
    evalCtx: IRBuilder,
    eltR: TypeWithRequiredness,
  ): IR = {
    val nd = s.collectLocal(evalCtx, "block_matrix_binary_writer")

    // FIXME remove numpy encoder
    val etype = ENumpyBinaryNDArray(s.typ.nRows, s.typ.nCols, true)
    val spec = TypedCodecSpec(etype, TNDArray(s.typ.elementType, Nat(2)), new StreamBufferSpec())
    val writer = ETypeValueWriter(spec)
    WriteValue(nd, Str(path), writer)
  }
}

case class BlockMatrixPersistWriter(id: String, storageLevel: String) extends BlockMatrixWriter {
  def pathOpt: Option[String] = None

  def apply(ctx: ExecuteContext, bm: BlockMatrix): Unit =
    HailContext.backend.persist(ctx.backendContext, id, bm, storageLevel)

  def loweredTyp: Type = TVoid
}

case class BlockMatrixRectanglesWriter(
  path: String,
  rectangles: Array[Array[Long]],
  delimiter: String,
  binary: Boolean,
) extends BlockMatrixWriter {

  def pathOpt: Option[String] = Some(path)

  def apply(ctx: ExecuteContext, bm: BlockMatrix): Unit =
    bm.exportRectangles(ctx, path, rectangles, delimiter, binary)

  def loweredTyp: Type = TVoid
}

abstract class BlockMatrixMultiWriter {
  def apply(ctx: ExecuteContext, bms: IndexedSeq[BlockMatrix]): Unit
}

case class BlockMatrixBinaryMultiWriter(
  prefix: String,
  overwrite: Boolean,
) extends BlockMatrixMultiWriter {

  def apply(ctx: ExecuteContext, bms: IndexedSeq[BlockMatrix]): Unit =
    BlockMatrix.binaryWriteBlockMatrices(ctx.fs, bms, prefix, overwrite)

  def loweredTyp: Type = TVoid
}

case class BlockMatrixTextMultiWriter(
  prefix: String,
  overwrite: Boolean,
  delimiter: String,
  header: Option[String],
  addIndex: Boolean,
  compression: Option[String],
  customFilenames: Option[Array[String]],
) extends BlockMatrixMultiWriter {

  def apply(ctx: ExecuteContext, bms: IndexedSeq[BlockMatrix]): Unit =
    BlockMatrix.exportBlockMatrices(
      ctx.fs,
      bms,
      prefix,
      overwrite,
      delimiter,
      header,
      addIndex,
      compression,
      customFilenames,
    )

  def loweredTyp: Type = TVoid
}

case class BlockMatrixNativeMultiWriter(
  prefix: String,
  overwrite: Boolean,
  forceRowMajor: Boolean,
) extends BlockMatrixMultiWriter {

  def apply(ctx: ExecuteContext, bms: IndexedSeq[BlockMatrix]): Unit =
    BlockMatrix.writeBlockMatrices(ctx, bms, prefix, overwrite, forceRowMajor)

  def loweredTyp: Type = TVoid
}
